File: bindings/php/SortStrategy.php
<?php
namespace Stencil;

final class SortStrategy {
    /** @internal */
    public \FFI\CData $ptr;
    private bool $owned;

    /** @internal */
    public function __construct(\FFI\CData $ptr, bool $owned) {
        $this->ptr = $ptr;
        $this->owned = $owned;
    }

    public static function yxz() {
        $ret = Lib::ffi()->SortStrategy_yxz();
        return new SortStrategy($ret, true);
    }

    public static function xyz() {
        $ret = Lib::ffi()->SortStrategy_xyz();
        return new SortStrategy($ret, true);
    }

    public static function zyx() {
        $ret = Lib::ffi()->SortStrategy_zyx();
        return new SortStrategy($ret, true);
    }

    public static function yDescXz() {
        $ret = Lib::ffi()->SortStrategy_y_desc_xz();
        return new SortStrategy($ret, true);
    }

    public static function xDescYz() {
        $ret = Lib::ffi()->SortStrategy_x_desc_yz();
        return new SortStrategy($ret, true);
    }

    public static function zDescYx() {
        $ret = Lib::ffi()->SortStrategy_z_desc_yx();
        return new SortStrategy($ret, true);
    }

    public static function descending() {
        $ret = Lib::ffi()->SortStrategy_descending();
        return new SortStrategy($ret, true);
    }

    public static function distanceFrom( $x,  $y,  $z) {
        $ret = Lib::ffi()->SortStrategy_distance_from($x, $y, $z);
        return new SortStrategy($ret, true);
    }

    public static function distanceFromDesc( $x,  $y,  $z) {
        $ret = Lib::ffi()->SortStrategy_distance_from_desc($x, $y, $z);
        return new SortStrategy($ret, true);
    }

    public static function preserve() {
        $ret = Lib::ffi()->SortStrategy_preserve();
        return new SortStrategy($ret, true);
    }

    public static function reverse() {
        $ret = Lib::ffi()->SortStrategy_reverse();
        return new SortStrategy($ret, true);
    }

    public static function fromString(string $s) {
        $__n0 = strlen($s);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $s, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $result = Lib::ffi()->SortStrategy_from_string($__view0);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return new SortStrategy($result->ok, true);
    }

    public function name() {
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        Lib::ffi()->SortStrategy_name($this->ptr, $write);
        return Lib::readAndFreeWrite($write);
    }

    public function __destruct() {
        if ($this->owned) {
            Lib::ffi()->SortStrategy_destroy($this->ptr);
        }
    }
}

File: bindings/kotlin/src/main/kotlin/at/schem/nucleation/SortStrategy.kt
package at.schem.nucleation;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface SortStrategyLib: Library {
    fun SortStrategy_destroy(handle: Pointer)
    fun SortStrategy_yxz(): Pointer
    fun SortStrategy_xyz(): Pointer
    fun SortStrategy_zyx(): Pointer
    fun SortStrategy_y_desc_xz(): Pointer
    fun SortStrategy_x_desc_yz(): Pointer
    fun SortStrategy_z_desc_yx(): Pointer
    fun SortStrategy_descending(): Pointer
    fun SortStrategy_distance_from(x: Int, y: Int, z: Int): Pointer
    fun SortStrategy_distance_from_desc(x: Int, y: Int, z: Int): Pointer
    fun SortStrategy_preserve(): Pointer
    fun SortStrategy_reverse(): Pointer
    fun SortStrategy_from_string(s: Slice): ResultPointerInt
    fun SortStrategy_name(handle: Pointer, write: Pointer): Unit
}
/** Ordering applied to region positions before bit assignment
*(PORTING rule 10).
*/
class SortStrategy internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal var owned: Boolean,
)  {

    init {
        if (this.owned) {
            this.registerCleaner()
        }
    }

    private class SortStrategyCleaner(val handle: Pointer, val lib: SortStrategyLib) : Runnable {
        override fun run() {
            lib.SortStrategy_destroy(handle)
        }
    }
    private fun registerCleaner() {
        CLEANER.register(this, SortStrategy.SortStrategyCleaner(handle, SortStrategy.lib));
    }

    companion object {
        internal val libClass: Class<SortStrategyLib> = SortStrategyLib::class.java
        internal val lib: SortStrategyLib = Native.load("nucleation", libClass)
        @JvmStatic
        
        fun yxz(): SortStrategy {
            
            val returnVal = lib.SortStrategy_yxz();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = SortStrategy(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun xyz(): SortStrategy {
            
            val returnVal = lib.SortStrategy_xyz();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = SortStrategy(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun zyx(): SortStrategy {
            
            val returnVal = lib.SortStrategy_zyx();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = SortStrategy(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun yDescXz(): SortStrategy {
            
            val returnVal = lib.SortStrategy_y_desc_xz();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = SortStrategy(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun xDescYz(): SortStrategy {
            
            val returnVal = lib.SortStrategy_x_desc_yz();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = SortStrategy(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun zDescYx(): SortStrategy {
            
            val returnVal = lib.SortStrategy_z_desc_yx();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = SortStrategy(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        /** Old ABI name: `sort_strategy_descending`.
        */
        fun descending(): SortStrategy {
            
            val returnVal = lib.SortStrategy_descending();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = SortStrategy(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun distanceFrom(x: Int, y: Int, z: Int): SortStrategy {
            
            val returnVal = lib.SortStrategy_distance_from(x, y, z);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = SortStrategy(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun distanceFromDesc(x: Int, y: Int, z: Int): SortStrategy {
            
            val returnVal = lib.SortStrategy_distance_from_desc(x, y, z);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = SortStrategy(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun preserve(): SortStrategy {
            
            val returnVal = lib.SortStrategy_preserve();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = SortStrategy(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun reverse(): SortStrategy {
            
            val returnVal = lib.SortStrategy_reverse();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = SortStrategy(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        /** Parse from a name (e.g. "yxz", "descending", "preserve").
        */
        fun fromString(s: String): Result<SortStrategy> {
            val sSliceMemory = PrimitiveArrayTools.borrowUtf8(s)
            
            val returnVal = lib.SortStrategy_from_string(sSliceMemory.slice);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    val selfEdges: List<Any> = listOf()
                    val handle = nativeOkVal 
                    val returnOpaque = SortStrategy(handle, selfEdges, true)
                    return returnOpaque.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                sSliceMemory.close()
            }
        }
    }
    
    /** The strategy name.
    */
    fun name(): String {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.SortStrategy_name(handle, write);
        
        val returnString = DW.writeToString(write)
        return returnString
    }

}